api/clients/eigenda_client_v2.go

```diff
@@ -0,0 +1,148 @@
+package clients
+
+import (
+	"context"
+	"fmt"
+	"github.com/Layr-Labs/eigenda/api/clients/codecs"
+	corev2 "github.com/Layr-Labs/eigenda/core/v2"
+	"github.com/Layr-Labs/eigensdk-go/logging"
+	"github.com/cockroachdb/errors/join"
+	"math/rand"
+)
+
+// EigenDAClientV2 provides the ability to get blobs from the relay subsystem, and to send new blobs to the disperser.
+type EigenDAClientV2 struct {
+	log logging.Logger
+	// doesn't need to be cryptographically secure, as it's only used to distribute load across relays
+	random      *rand.Rand
+	config      *EigenDAClientConfigV2
+	codec       codecs.BlobCodec
+	relayClient RelayClient
+}
+
+// BuildEigenDAClientV2 builds an EigenDAClientV2 from config structs.
+func BuildEigenDAClientV2(
+	log logging.Logger,
+	config *EigenDAClientConfigV2,
+	relayClientConfig *RelayClientConfig) (*EigenDAClientV2, error) {
+
+	relayClient, err := NewRelayClient(relayClientConfig, log)
+	if err != nil {
+		return nil, fmt.Errorf("new relay client: %w", err)
+	}
+
+	codec, err := createCodec(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewEigenDAClientV2(log, rand.New(rand.NewSource(rand.Int63())), config, relayClient, codec)
+}
+
+// NewEigenDAClientV2 assembles an EigenDAClientV2 from subcomponents that have already been constructed and initialized.
+func NewEigenDAClientV2(
+	log logging.Logger,
+	random *rand.Rand,
+	config *EigenDAClientConfigV2,
+	relayClient RelayClient,
+	codec codecs.BlobCodec) (*EigenDAClientV2, error) {
+
+	return &EigenDAClientV2{
+		log:         log,
+		random:      random,
+		config:      config,
+		codec:       codec,
+		relayClient: relayClient,
+	}, nil
+}
+
+// GetBlob iteratively attempts to retrieve a given blob with key blobKey from the relays listed in the blobCertificate.
+//
+// The relays are attempted in random order.
+//
+// The returned blob is decoded.
+func (c *EigenDAClientV2) GetBlob(
+	ctx context.Context,
+	blobKey corev2.BlobKey,
+	blobCertificate corev2.BlobCertificate) ([]byte, error) {
+
+	relayKeyCount := len(blobCertificate.RelayKeys)
+
+	if relayKeyCount == 0 {
+		return nil, fmt.Errorf("relay key count is zero")
+	}
+
+	var indices []int
+	// create a randomized array of indices, so that it isn't always the first relay in the list which gets hit
+	for i := 0; i < relayKeyCount; i++ {
+		indices = append(indices, i)
+	}
```

Comment:
nit:
```
indices := make([]int, relayKeyCount)
for i := 0; i < relayKeyCount; i++ {
  indices[i] = i
}
```

Reply:
I simplified even further, as suggested in [this comment](https://github.com/Layr-Labs/eigenda/pull/972#discussion_r1880757891)

Reply:
I simplified even further, as suggested in [this comment](https://github.com/Layr-Labs/eigenda/pull/972#discussion_r1880757891)

[a48afb1](https://github.com/Layr-Labs/eigenda/pull/972/commits/a48afb152a4bf90cc901c2f1ba6bfcd23cf90cca)